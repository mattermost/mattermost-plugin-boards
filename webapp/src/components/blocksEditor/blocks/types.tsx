// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export type BlockInputProps<ValueType = string> = {
    onChange: (value: ValueType) => void
    value: ValueType
    onCancel: () => void
    onSave: (val: ValueType) => void
    currentBoardId?: string
}

export type ContentType<ValueType = string> = {
    name: string
    displayName: string
    slashCommand: string
    prefix: string
    editable: boolean
    Input: React.FunctionComponent<BlockInputProps<ValueType>>
    Display: React.FunctionComponent<BlockInputProps<ValueType>>
    runSlashCommand: (changeType: (contentType: ContentType<ValueType>) => void, changeValue: (value: ValueType) => void, ...args: string[]) => void
    nextType?: string
}

export type BlockData<ValueType = string> = {
    id?: string
    value: ValueType
    contentType: string
}
