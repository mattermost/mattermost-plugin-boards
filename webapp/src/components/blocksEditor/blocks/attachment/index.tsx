// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useRef, useEffect, useState} from 'react'

import {BlockInputProps, ContentType} from '../types'
import octoClient from '../../../../octoClient'

import './attachment.scss'

type FileInfo = {
    file: string|File
    filename: string
}

const Attachment: ContentType<FileInfo> = {
    name: 'attachment',
    displayName: 'Attachment',
    slashCommand: '/attachment',
    prefix: '',
    runSlashCommand: (): void => {},
    editable: false,
    Display: (props: BlockInputProps<FileInfo>) => {
        const [fileDataUrl, setFileDataUrl] = useState<string|null>(null)

        useEffect(() => {
            if (!fileDataUrl) {
                const loadFile = async () => {
                    if (props.value && props.value.file && typeof props.value.file === 'string') {
                        const fileURL = await octoClient.getFileAsDataUrl(props.currentBoardId || '', props.value.file)
                        setFileDataUrl(fileURL.url || '')
                    }
                }
                loadFile()
            }
        }, [props.value, props.value.file, props.currentBoardId])

        return (
            <div
                className='AttachmentView'
                data-testid='attachment'
            >
                <a
                    href={fileDataUrl || '#'}
                    onClick={(e) => e.stopPropagation()}
                    download={props.value.filename}
                >
                    {'ðŸ“Ž'} {props.value.filename}
                </a>
            </div>
        )
    },
    Input: (props: BlockInputProps<FileInfo>) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <input
                ref={ref}
                className='Attachment'
                data-testid='attachment-input'
                type='file'
                onChange={(e) => {
                    const files = e.currentTarget?.files
                    if (files) {
                        for (let i = 0; i < files.length; i++) {
                            const file = files.item(i)
                            if (file) {
                                props.onSave({file, filename: file.name})
                            }
                        }
                    }
                }}
            />
        )
    },
}

Attachment.runSlashCommand = (changeType: (contentType: ContentType<FileInfo>) => void, changeValue: (value: FileInfo) => void): void => {
    changeType(Attachment)
    changeValue({} as any)
}

export default Attachment
