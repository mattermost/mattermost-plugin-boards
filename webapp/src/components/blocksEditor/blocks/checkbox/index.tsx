// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './checkbox.scss'

type ValueType = {
    value: string
    checked: boolean
}

const Checkbox: ContentType<ValueType> = {
    name: 'checkbox',
    displayName: 'Checkbox',
    slashCommand: '/checkbox',
    prefix: '[ ] ',
    nextType: 'checkbox',
    runSlashCommand: (): void => {},
    editable: true,
    Display: (props: BlockInputProps<ValueType>) => {
        const renderer = new marked.Renderer()
        const html = marked(props.value.value || '', {renderer, breaks: true})
        return (
            <div className='CheckboxView'>
                <input
                    data-testid='checkbox-check'
                    type='checkbox'
                    onChange={(e) => {
                        const newValue = {checked: Boolean(e.target.checked), value: props.value.value || ''}
                        props.onSave(newValue)
                    }}
                    checked={props.value.checked || false}
                    onClick={(e) => e.stopPropagation()}
                />
                <div
                    dangerouslySetInnerHTML={{__html: html.trim()}}
                />
            </div>
        )
    },
    Input: (props: BlockInputProps<ValueType>) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <div className='Checkbox'>
                <input
                    type='checkbox'
                    data-testid='checkbox-check'
                    className='inputCheck'
                    onChange={(e) => {
                        let newValue = {checked: false, value: props.value.value || ''}
                        if (e.target.checked) {
                            newValue = {checked: true, value: props.value.value || ''}
                        }
                        props.onChange(newValue)
                        ref.current?.focus()
                    }}
                    checked={props.value.checked || false}
                />
                <input
                    ref={ref}
                    data-testid='checkbox-input'
                    className='inputText'
                    onChange={(e) => {
                        props.onChange({checked: Boolean(props.value.checked), value: e.currentTarget.value})
                    }}
                    onKeyDown={(e) => {
                        if ((props.value.value || '') === '' && e.key === 'Backspace') {
                            props.onCancel()
                        }
                        if (e.key === 'Enter') {
                            props.onSave(props.value || {checked: false, value: ''})
                        }
                    }}
                    value={props.value.value || ''}
                />
            </div>
        )
    },
}

Checkbox.runSlashCommand = (changeType: (contentType: ContentType<ValueType>) => void, changeValue: (value: ValueType) => void, ...args: string[]): void => {
    changeType(Checkbox)
    changeValue({checked: false, value: args.join(' ')})
}

export default Checkbox
