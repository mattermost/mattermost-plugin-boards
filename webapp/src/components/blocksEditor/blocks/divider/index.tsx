// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useEffect} from 'react'

import {BlockInputProps, ContentType} from '../types'

import './divider.scss'

const Divider: ContentType = {
    name: 'divider',
    displayName: 'Divider',
    slashCommand: '/divider',
    prefix: '--- ',
    runSlashCommand: (): void => {},
    editable: false,
    Display: () => <hr className='Divider'/>,
    Input: (props: BlockInputProps) => {
        useEffect(() => {
            props.onSave(props.value)
        }, [])
        return null
    },
}

Divider.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(Divider)
    changeValue(args.join(' '))
}

export default Divider
