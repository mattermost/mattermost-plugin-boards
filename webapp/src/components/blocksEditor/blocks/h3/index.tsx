// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useRef, useEffect} from 'react'
import {marked} from 'marked'

import {BlockInputProps, ContentType} from '../types'

import './h3.scss'

const H3: ContentType = {
    name: 'h3',
    displayName: 'Sub Sub title',
    slashCommand: '/subsubtitle',
    prefix: '### ',
    runSlashCommand: (): void => {},
    editable: true,
    Display: (props: BlockInputProps) => {
        const renderer = new marked.Renderer()
        const html = marked('### ' + props.value, {renderer, breaks: true})
        return (
            <div
                dangerouslySetInnerHTML={{__html: html.trim()}}
            />
        )
    },
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <input
                ref={ref}
                className='H3'
                data-testid='h3'
                onChange={(e) => props.onChange(e.currentTarget.value)}
                onKeyDown={(e) => {
                    if (props.value === '' && e.key === 'Backspace') {
                        props.onCancel()
                    }
                    if (e.key === 'Enter') {
                        props.onSave(props.value)
                    }
                }}
                value={props.value}
            />
        )
    },
}

H3.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(H3)
    changeValue(args.join(' '))
}

export default H3
