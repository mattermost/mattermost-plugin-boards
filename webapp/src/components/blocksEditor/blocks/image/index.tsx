// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useRef, useEffect, useState} from 'react'

import {BlockInputProps, ContentType} from '../types'
import octoClient from '../../../../octoClient'

import './image.scss'

type FileInfo = {
    file: string|File
    width?: number
    align?: 'left'|'center'|'right'
}

const Image: ContentType<FileInfo> = {
    name: 'image',
    displayName: 'Image',
    slashCommand: '/image',
    prefix: '',
    runSlashCommand: (): void => {},
    editable: false,
    Display: (props: BlockInputProps<FileInfo>) => {
        const [imageDataUrl, setImageDataUrl] = useState<string|null>(null)

        useEffect(() => {
            if (!imageDataUrl) {
                const loadImage = async () => {
                    if (props.value && props.value.file && typeof props.value.file === 'string') {
                        const fileURL = await octoClient.getFileAsDataUrl(props.currentBoardId || '', props.value.file)
                        setImageDataUrl(fileURL.url || '')
                    }
                }
                loadImage()
            }
        }, [props.value, props.value.file, props.currentBoardId])

        if (imageDataUrl) {
            return (
                <img
                    data-testid='image'
                    className='ImageView'
                    src={imageDataUrl}
                />
            )
        }
        return null
    },
    Input: (props: BlockInputProps<FileInfo>) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.click()
        }, [])

        return (
            <div>
                {props.value.file && (typeof props.value.file === 'string') && (
                    <img
                        className='ImageView'
                        src={props.value.file}
                        onClick={() => ref.current?.click()}
                    />
                )}
                <input
                    ref={ref}
                    className='Image'
                    data-testid='image-input'
                    type='file'
                    accept='image/*'
                    onChange={(e) => {
                        const file = (e.currentTarget?.files || [])[0]
                        props.onSave({file})
                    }}
                />
            </div>
        )
    },
}

Image.runSlashCommand = (changeType: (contentType: ContentType<FileInfo>) => void, changeValue: (value: FileInfo) => void): void => {
    changeType(Image)
    changeValue({file: ''})
}

export default Image
