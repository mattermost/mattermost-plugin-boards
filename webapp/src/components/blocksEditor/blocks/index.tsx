// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {ContentType} from './types'
import H1 from './h1'
import H2 from './h2'
import H3 from './h3'
import Image from './image'
import Text from './text'
import Divider from './divider'

// import Markdown from './markdown'
import ListItem from './list-item'
import Attachment from './attachment'
import Quote from './quote'
import Video from './video'
import Checkbox from './checkbox'

const blocks: {[key: string]: ContentType} = {}
const blocksByPrefix: {[key: string]: ContentType} = {}
const blocksBySlashCommand: {[key: string]: ContentType} = {}

export function register(contentType: ContentType<any>) {
    blocks[contentType.name] = contentType
    if (contentType.prefix !== '') {
        blocksByPrefix[contentType.prefix] = contentType
    }
    blocksBySlashCommand[contentType.slashCommand] = contentType
}

export function list() {
    return Object.values(blocks)
}

export function get(name: string): ContentType {
    return blocks[name]
}

export function getByPrefix(prefix: string): ContentType {
    return blocksByPrefix[prefix]
}

export function isSubPrefix(text: string): boolean {
    for (const ct of list()) {
        if (ct.prefix !== '' && ct.prefix.startsWith(text)) {
            return true
        }
    }
    return false
}

export function getBySlashCommand(slashCommand: string): ContentType {
    return blocksBySlashCommand[slashCommand]
}

export function getBySlashCommandPrefix(slashCommandPrefix: string): ContentType|null {
    for (const ct of list()) {
        if (ct.slashCommand.startsWith(slashCommandPrefix)) {
            return ct
        }
    }
    return null
}

register(H1)
register(H2)
register(H3)
register(Image)
register(Text)
register(Divider)

// register(Markdown)
register(ListItem)
register(Attachment)
register(Quote)
register(Video)
register(Checkbox)
