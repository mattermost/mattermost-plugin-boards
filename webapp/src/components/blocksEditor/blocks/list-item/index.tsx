// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useRef, useEffect} from 'react'

import {BlockInputProps, ContentType} from '../types'

import './list-item.scss'

const ListItem: ContentType = {
    name: 'list-item',
    displayName: 'List item',
    slashCommand: '/list-item',
    prefix: '* ',
    nextType: 'list-item',
    runSlashCommand: (): void => {},
    editable: true,
    Display: (props: BlockInputProps) => <ul><li>{props.value}</li></ul>,
    Input: (props: BlockInputProps) => {
        const ref = useRef<HTMLInputElement|null>(null)
        useEffect(() => {
            ref.current?.focus()
        }, [])
        return (
            <ul>
                <li>
                    <input
                        ref={ref}
                        className='ListItem'
                        data-testid='list-item'
                        onChange={(e) => props.onChange(e.currentTarget.value)}
                        onKeyDown={(e) => {
                            if (props.value === '' && e.key === 'Backspace') {
                                props.onCancel()
                            }
                            if (e.key === 'Enter') {
                                props.onSave(props.value)
                            }
                        }}
                        value={props.value}
                    />
                </li>
            </ul>
        )
    },
}

ListItem.runSlashCommand = (changeType: (contentType: ContentType) => void, changeValue: (value: string) => void, ...args: string[]): void => {
    changeType(ListItem)
    changeValue(args.join(' '))
}

export default ListItem
