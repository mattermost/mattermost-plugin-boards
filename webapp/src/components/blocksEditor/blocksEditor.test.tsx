// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.


import React from 'react'
import {Provider as ReduxProvider} from 'react-redux'
import {render, screen, fireEvent, act} from '@testing-library/react'

import {mockDOM, wrapDNDIntl, mockStateStore} from '../../testUtils'
import {TestBlockFactory} from '../../test/testBlockFactory'

import {BlockData} from './blocks/types'
import BlocksEditor from './blocksEditor'

jest.mock('draft-js/lib/generateRandomKey', () => () => '123')

describe('components/blocksEditor/blocksEditor', () => {
    beforeEach(mockDOM)

    const blocks: Array<BlockData<any>> = [
        {id: '1', value: 'Title', contentType: 'h1'},
        {id: '2', value: 'Sub title', contentType: 'h2'},
        {id: '3', value: 'Sub sub title', contentType: 'h3'},
        {id: '4', value: 'Some **markdown** text', contentType: 'text'},
        {id: '5', value: 'Some multiline\n**markdown** text\n### With Items\n- Item 1\n- Item2\n- Item3', contentType: 'text'},
        {id: '6', value: {checked: true, value: 'Checkbox'}, contentType: 'checkbox'},
    ]

    const board1 = TestBlockFactory.createBoard()
    board1.id = 'board-id-1'

    const state = {
        users: {
            boardUsers: {
                1: {username: 'abc'},
                2: {username: 'd'},
                3: {username: 'e'},
                4: {username: 'f'},
                5: {username: 'g'},
            },
        },
        boards: {
            current: 'board-id-1',
            boards: {
                [board1.id]: board1,
            },
        },
        clientConfig: {
            value: {},
        },
    }
    const store = mockStateStore([], state)

    test('should match snapshot on empty', async () => {
        let container
        await act(async () => {
            const result = render(wrapDNDIntl(
                <ReduxProvider store={store}>
                    <BlocksEditor
                        boardId='test-board'
                        onBlockCreated={jest.fn()}
                        onBlockModified={jest.fn()}
                        onBlockMoved={jest.fn()}
                        blocks={[]}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should match snapshot with blocks', async () => {
        let container
        await act(async () => {
            const result = render(wrapDNDIntl(
                <ReduxProvider store={store}>
                    <BlocksEditor
                        boardId='test-board'
                        onBlockCreated={jest.fn()}
                        onBlockModified={jest.fn()}
                        onBlockMoved={jest.fn()}
                        blocks={blocks}
                    />
                </ReduxProvider>,
            ))
            container = result.container
        })
        expect(container).toMatchSnapshot()
    })

    test('should call onBlockCreate after introduce text and hit enter', async () => {
        const onBlockCreated = jest.fn()
        await act(async () => {
            render(wrapDNDIntl(
                <ReduxProvider store={store}>
                    <BlocksEditor
                        boardId='test-board'
                        onBlockCreated={onBlockCreated}
                        onBlockModified={jest.fn()}
                        onBlockMoved={jest.fn()}
                        blocks={[]}
                    />
                </ReduxProvider>,
            ))
        })

        let input = screen.getByDisplayValue('')
        expect(onBlockCreated).not.toBeCalled()
        fireEvent.change(input, {target: {value: '/title'}})
        fireEvent.keyDown(input, {key: 'Enter'})

        input = screen.getByDisplayValue('')
        fireEvent.change(input, {target: {value: 'test'}})
        fireEvent.keyDown(input, {key: 'Enter'})

        expect(onBlockCreated).toBeCalledWith(expect.objectContaining({value: 'test'}))
    })

    test('should call onBlockModified after introduce text and hit enter', async () => {
        const onBlockModified = jest.fn()
        await act(async () => {
            render(wrapDNDIntl(
                <ReduxProvider store={store}>
                    <BlocksEditor
                        boardId='test-board'
                        onBlockCreated={jest.fn()}
                        onBlockModified={onBlockModified}
                        onBlockMoved={jest.fn()}
                        blocks={blocks}
                    />
                </ReduxProvider>,
            ))
            const input = screen.getByTestId('checkbox-check')
            expect(onBlockModified).not.toBeCalled()
            fireEvent.click(input)
            expect(onBlockModified).toBeCalledWith(expect.objectContaining({value: {checked: false, value: 'Checkbox'}}))
        })
    })
})
