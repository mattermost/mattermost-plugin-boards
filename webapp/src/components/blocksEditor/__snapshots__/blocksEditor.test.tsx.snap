// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/blocksEditor/blocksEditor should match snapshot on empty 1`] = `
<div>
  <div
    class="BlocksEditor"
  >
    <div
      class="Editor"
    >
      <div
        class="RootInput css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-2-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-1haocjs-control"
        >
          <div
            class=" css-b2z5qd-ValueContainer"
          >
            <div
              class=" css-1jqq78o-placeholder"
              id="react-select-2-placeholder"
            >
              Introduce your text or your slash command
            </div>
            <div
              class=" css-26cneq-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-2-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-2-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components/blocksEditor/blocksEditor should match snapshot with blocks 1`] = `
<div>
  <div
    class="BlocksEditor"
  >
    <div>
      <div
        class="BlockContent   "
        data-testid="block-content"
        style="opacity: 1;"
      >
        <span
          class="action"
          data-testid="add-action"
        >
          <i
            class="CompassIcon icon-plus AddIcon"
          />
        </span>
        <span
          class="action"
          draggable="true"
        >
          <svg
            class="GripIcon Icon"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M11 18c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm-2-8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
            />
          </svg>
        </span>
        <div
          class="content"
        >
          <div>
            <h1
              id="title"
            >
              Title
            </h1>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="BlockContent   "
        data-testid="block-content"
        style="opacity: 1;"
      >
        <span
          class="action"
          data-testid="add-action"
        >
          <i
            class="CompassIcon icon-plus AddIcon"
          />
        </span>
        <span
          class="action"
          draggable="true"
        >
          <svg
            class="GripIcon Icon"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M11 18c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm-2-8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
            />
          </svg>
        </span>
        <div
          class="content"
        >
          <div>
            <h2
              id="sub-title"
            >
              Sub title
            </h2>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="BlockContent   "
        data-testid="block-content"
        style="opacity: 1;"
      >
        <span
          class="action"
          data-testid="add-action"
        >
          <i
            class="CompassIcon icon-plus AddIcon"
          />
        </span>
        <span
          class="action"
          draggable="true"
        >
          <svg
            class="GripIcon Icon"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M11 18c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm-2-8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
            />
          </svg>
        </span>
        <div
          class="content"
        >
          <div>
            <h3
              id="sub-sub-title"
            >
              Sub sub title
            </h3>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="BlockContent   "
        data-testid="block-content"
        style="opacity: 1;"
      >
        <span
          class="action"
          data-testid="add-action"
        >
          <i
            class="CompassIcon icon-plus AddIcon"
          />
        </span>
        <span
          class="action"
          draggable="true"
        >
          <svg
            class="GripIcon Icon"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M11 18c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm-2-8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
            />
          </svg>
        </span>
        <div
          class="content"
        >
          <div
            class="octo-editor-preview"
          >
            <p>
              Some 
              <strong>
                markdown
              </strong>
               text
            </p>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="BlockContent   "
        data-testid="block-content"
        style="opacity: 1;"
      >
        <span
          class="action"
          data-testid="add-action"
        >
          <i
            class="CompassIcon icon-plus AddIcon"
          />
        </span>
        <span
          class="action"
          draggable="true"
        >
          <svg
            class="GripIcon Icon"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M11 18c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm-2-8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
            />
          </svg>
        </span>
        <div
          class="content"
        >
          <div
            class="octo-editor-preview"
          >
            <p>
              Some multiline
              <br />
              <strong>
                markdown
              </strong>
               text
            </p>
            

            <h3
              id="with-items"
            >
              With Items
            </h3>
            

            <ul>
              

              <li>
                Item 1
              </li>
              

              <li>
                Item2
              </li>
              

              <li>
                Item3
              </li>
              

            </ul>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="BlockContent   "
        data-testid="block-content"
        style="opacity: 1;"
      >
        <span
          class="action"
          data-testid="add-action"
        >
          <i
            class="CompassIcon icon-plus AddIcon"
          />
        </span>
        <span
          class="action"
          draggable="true"
        >
          <svg
            class="GripIcon Icon"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M11 18c0 1.1-.9 2-2 2s-2-.9-2-2 .9-2 2-2 2 .9 2 2zm-2-8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0-6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 4c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
            />
          </svg>
        </span>
        <div
          class="content"
        >
          <div
            class="CheckboxView"
          >
            <input
              checked=""
              data-testid="checkbox-check"
              type="checkbox"
            />
            <div>
              <p>
                Checkbox
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="Editor"
    >
      <div
        class="RootInput css-b62m3t-container"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-3-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
        />
        <div
          class=" css-1haocjs-control"
        >
          <div
            class=" css-b2z5qd-ValueContainer"
          >
            <div
              class=" css-1jqq78o-placeholder"
              id="react-select-3-placeholder"
            >
              Introduce your text or your slash command
            </div>
            <div
              class=" css-26cneq-Input"
              data-value=""
            >
              <input
                aria-autocomplete="list"
                aria-describedby="react-select-3-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-3-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
